File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/myshop/api/SystemErrorMapper.java
package com.mycompany.myshop.testkit.driver.adapter.myshop.api;

import com.mycompany.myshop.testkit.driver.port.myshop.dtos.error.SystemError;
import com.mycompany.myshop.testkit.driver.adapter.myshop.api.client.dtos.errors.ProblemDetailResponse;

public class SystemErrorMapper {
    private SystemErrorMapper() {
    }

    public static SystemError from(ProblemDetailResponse problemDetail) {
        var message = problemDetail.getDetail() != null ? problemDetail.getDetail() : "Request failed";

        if (problemDetail.getErrors() != null && !problemDetail.getErrors().isEmpty()) {
            var fieldErrors = problemDetail.getErrors().stream()
                    .map(e -> new SystemError.FieldError(e.getField(), e.getMessage(), e.getCode()))
                    .toList();
            return SystemError.of(message, fieldErrors);
        }

        return SystemError.of(message);
    }
}

File: system-test/dotnet/Driver.Adapter/MyShop/Api/SystemErrorMapper.cs
using Driver.Adapter.MyShop.Api.Client.Dtos.Errors;
using Driver.Port.MyShop.Dtos.Error;

namespace Driver.Adapter.MyShop.Api;

public static class SystemErrorMapper
{
    public static SystemError From(ProblemDetailResponse problemDetail)
    {
        var message = problemDetail.Detail ?? "Request failed";

        if (problemDetail.Errors != null && problemDetail.Errors.Count > 0)
        {
            var fieldErrors = problemDetail.Errors
                .Select(e => new SystemError.FieldError(e.Field ?? string.Empty, e.Message ?? string.Empty, e.Code))
                .ToList()
                .AsReadOnly();
            return SystemError.Of(message, fieldErrors);
        }

        return SystemError.Of(message);
    }
}
